easy-ui-react/src/TextField/TextField.tsx

```diff
@@ -0,0 +1,359 @@
+ *        value={password}
+ *        onChange={(inputValue) => setPassword(inputValue)} // value is returned automatically via react-aria
+ *        isRequired
+ *      />
+ *    </>
+ *  );
+ * }
+ * ```
+ * @example
+ * _Visually hidden label with left aligned icon and placeholder text:_
+ * ```tsx
+ * import { TextField } from "@easypost/easy-ui/TextField";
+ *
+ * export function Component() {
+ *  const [searchedValue, setSearchedValue] = useState("");
+ *  return (
+ *    <>
+ *      <TextField
+ *        type="search"
+ *        label="Search for carriers" // visually hidden but still accessible via isLabelVisuallyHidden prop
+ *        isLabelVisuallyHidden
+ *        iconAtStart={SearchIcon}
+ *        value={searchedValue}
+ *        onChange={(inputValue) => setSearchedValue(inputValue)} // value is returned automatically via react-aria
+ *        placeholder="FedEx, UPS, USPS"
+ *        isRequired
+ *      />
+ *    </>
+ *  );
+ * }
+ * ```
+ * @example
+ * _Invalid state with error text:_
+ * ```tsx
+ * import { TextField } from "@easypost/easy-ui/TextField";
+ *
+ * export function Component() {
+ *  const [value, setValue] = useState("");
+ *  return (
+ *    <>
+ *      <TextField
+ *        label="Error"
+ *        validationState="invalid"
+ *        helperText="Some text" // will be overriden in the presence of an invalid state with error text
+ *        errorText="Some error text"
+ *        value={value}
+ *        onChange={(inputValue) => setValue(inputValue)} // value is returned automatically via react-aria
+ *        isRequired
+ *      />
+ *    </>
+ *  );
+ * }
+ * ```
+ */
+export function TextField(props: TextFieldProps) {
+  const {
+    type = "text",
+    size = "md",
+    isLabelVisuallyHidden = false,
+    isDisabled = false,
+    isRequired = false,
+    validationState = "valid",
+    emphasizedLabel = false,
+    autoFocus = false,
+    label,
+    errorText,
+    helperText,
+    placeholder,
+    value,
+    defaultValue,
+    iconAtStart,
+    iconAtEnd,
+  } = props;
+  const [isVisible, setIsVisible] = useState(false);
+  const ref = React.useRef(null);
+
+  const {
+    labelProps,
+    inputProps,
+    descriptionProps: helperTextProps,
+    errorMessageProps: errorTextProps,
+  } = useTextField(props, ref);
+
+  const bothIconPropsDefined = iconAtEnd && iconAtStart;
+  if (bothIconPropsDefined) {
+    // eslint-disable-next-line no-console
+    console.warn("Cannot simultaneously define `iconAtEnd` and `iconAtStart`");
+  }
+
+  const isPassword = type === "password";
+  const hasError = validationState === "invalid";
+  const showErrorText = hasError && errorText;
+  const showHelperText = !showErrorText && helperText;
+  const canUseIcon = !bothIconPropsDefined && !isPassword;
+  const hasStartIcon = canUseIcon && iconAtStart;
+  const hasEndIcon = canUseIcon && iconAtEnd;
+  const typeAdjustedForPasswordVisibility = isPassword && isVisible;
```

Comment:
**nit:** since this is a boolean could we name it `isTypeAdjustedForPasswordVisibility`?